newdocs/typeclass-cheatsheet.md

```diff
@@ -0,0 +1,223 @@
+   getHeader          :: b → Header b             -- extract header from the block
+   blockMatchesHeader :: Header b → b → Bool      -- check if the header is the header of the block
+   headerIsEBB        :: Header b → Maybe EpochNo -- when the header of an Epoch Boundary Block (EBB), ...
+ 
+ class (HasHeader b, GetHeader b) => GetPrevHash b where   
+   headerPrevHash :: Header b → ChainHash b       -- get the hash of predecessor
+ 
+ -- construct the two views on block 'b' required by protocol 'p'
+ class (GetPrevHash b, ConsensusProtocol p) => BlockSupportsProtocol b where              
+   validateView :: bc → Header b → ValidateView p  -- project from hdr for hdr validation
+   selectView   :: bc → Header b → SelectView p    -- project from hdr for chain selection
+```
+Classes connected to ledgers:
+```haskell
+  class GetTip l where                         
+    getTip :: l → Point l               -- Point of the most recently applied block
+
+  class (GetTip l, GetTip (Ticked l)) => IsLedger l where
+    type family LedgerErr l      :: Type                   
+    type family AuxLedgerEvent l :: Type
+    applyChainTickLedgerResult   :: lc → SlotNo → l → LedgerResult l (Ticked l)  -- apply slot based state transformations (tip unchanged)
+        
+  class (IsLedger l, HeaderHash l ~ HeaderHash b, HasHeader b, HasHeader (Header b)) => ApplyBlock l b where
+    applyBlockLedgerResult   :: lc → b → Ticked l → Except (LedgerErr l) (LedgerResult l l)  
+    reapplyBlockLedgerResult :: lc → b → Ticked l →                       LedgerResult l l
+    
+  class ApplyBlock (LedgerState b) b => UpdateLedger b where
+    {}
+
+  -- | Link protocol to ledger
+  class (BlockSupportsProtocol b, UpdateLedger b, ValidateEnvelope b) => LedgerSupportsProtocol b where
+    protocolLedgerView   :: lc → Ticked l → Ticked ledvw   -- 'ledvw' ('LedgerView (BlockProtocol b)') extracted from the ledger
+    ledgerViewForecastAt :: lc → l → Forecast ledvw        -- get a forecast (of future 'ledvw's) from a given ledger state.
+      
+  class (UpdateLedger b) => LedgerSupportsMempool b where
+    txInvariant :: GenTx b → Bool                                                -- check if internal invariants of the transaction hold
+    applyTx   :: lc → WhetherToIntervene → SlotNo → tx → tls → Except txerr (tls, Validated tx)      -- apply an unvalidated transaction
+    reapplyTx :: lc →            SlotNo → Validated tx → tls → Except txerr tls          -- apply a previously validated transaction ...
+    txsMaxBytes :: tls → Word32                                   -- max number of bytes of transactions that can be put into a block    
+    txInBlockSize :: tx → Word32                                  -- post-serialisation size in bytes of a 'GenTx b'
+    txForgetValidated :: Validated tx → tx                        -- discard the evidence that transaction has been previously validated     
+```
+
+## Quick Reference for Useful Base Types
```

Comment:
Initial reaction: I think the diagram is more helpful than the classes overview, and the classes overview is more helpful than this section of a excerpted declarations.

I figure that's an expected consequence of my years of experience on the project, and your more recent onboarding is probably the wiser influence. Moreover, the sections are already in my priority ordering, so 👍 

One action item, though: I suggest collapsing this section to be as vertically dense as the others. In particular, I'd drop the package and module name headers (rarely does this matter, and tags/HLS/grep should suffice when it does) and I'd try to thin out the bigger Haddock comments.

Reply:
> this section of a excerpted declarations.

Yes, even to me it feels ad hoc, just the selection of the types I had to look up a time or two, maybe this might evolve to a "better curated" set of types.  Will leave them though.

> One action item, though: I suggest collapsing this section ...

Done!